Support Scribble Handwriting

fbcouch opened the pull request:
Framework PR: 
https://github.com/flutter/flutter/pull/75472

This PR is the engine side of adding scribble handwriting support. I still need to clean up the NSLogs and add some tests, but looking for feedback on the overall approach. I did incorporate some feedback from @LongCatIsLooong on a previous draft, moving most of the logic of this into the framework and adding a widget there to handle most of the scribble-specific logic.

https://github.com/flutter/flutter/issues/61278

Design doc: https://docs.google.com/document/d/1mjQbsSRQnHuAgMNdouaSgTS-Xv-w57fdKfOUUafWpRo

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.
- [x] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

chunhtai commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1568,12 +1848,15 @@ - (void)resetAllClientIds {
-  if (![inputView isDescendantOfView:_inputHider]) {
-    [_inputHider addSubview:inputView];
-  }
+  // TODO: Find a different way to do what FlutterTextInputViewAccessibilityHider was doing
```
The reason we have this weird dangling views are our effort to try to make inputView harder to be reach, yet is still reachable.

If you directly attach to keywindow, it will be very easy to focus in voiceover by swipes left and right. For example, if you activate the textfield in ios, swipe left to move the focus around, and you will see the textinputview receives the focus (It will be redirect back to the flutter textfield, but it is still bad user experience )

fbcouch replied:
I think I was just doing something wrong in the merge – it seems to work fine now as a child of the inputHider

chinmaygarde commented:
@chunhtai This seems ready for review. Are you able to take a look please?

chunhtai commented on shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:
```diff
@@ -289,6 +289,8 @@ - (void)setViewController:(FlutterViewController*)viewController {
 
 - (void)attachView {
   self.iosPlatformView->attachView();
+  _textInputPlugin.get().viewController = [self viewController];
```
what if the view controller changes during the lifetime of the flutterengine?
also flutter holdsa weak ptr to the viewcontroller, I am not sure what will happen to the assign property of _textInputPlugin if the pointer is nulled in flutterengine, cc @gaaclarke

fbcouch replied:
@chunhtai I also set this property below on like 476, I'm not sure whether or not that would fix the issue – how would I test that? I don't know when the view controller would change

chunhtai replied:
I think the viewcontroller of the textinputplugin need to be set in FlutterEngine.setViewController, and it also need to nil it in FlutterEngine.notifyViewControllerDeallocated. To test this, you can swap out view controller when you run the FlutterEngine and see if things still work

fbcouch replied:
@chunhtai I may need some more guidance on how to test this...I tried the following, but the notifyViewControllerDeallocated method doesn't seem to get hit when swapping out the view controller – attachView does get called again, so I think it is "working":

```
@UIApplicationMain
@objc class AppDelegate: FlutterAppDelegate {
  override func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
  ) -> Bool {
    
    var controller = window.rootViewController as! FlutterViewController
    
    let criticChannel = FlutterMethodChannel(name: "test.viewController", binaryMessenger: controller as! FlutterBinaryMessenger)
    criticChannel.setMethodCallHandler { (call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in
        if call.method == "swap" {
            let engine = controller.engine!
            engine.viewController = nil
            self.window.rootViewController = FlutterViewController(engine: engine, nibName: controller.nibName, bundle: controller.nibBundle)
            controller = self.window.rootViewController as! FlutterViewController
            result(nil)
        } else {
            result(FlutterMethodNotImplemented)
        }
    }
    
    GeneratedPluginRegistrant.register(with: self)
    return super.application(application, didFinishLaunchingWithOptions: launchOptions)
  }
}
```

fbcouch replied:
@chunhtai I updated the `setupIndirectScribbleInteraction` method to take in the `FlutterViewController` instead of setting it directly in the engine, that way we're setting up that interaction every time the view controller changes. I also added a resetViewController method to nil the viewController from that `notifyViewControllerDeallocated` method, though I still haven't managed to actually have that get hit 🤷‍♂️

chunhtai replied:
you will need to hold a strong reference FlutterViewController somewhere and pass that into the flutter engine. and then nil the strong reference, this should cause garbage collection to dealloc the FlutterViewController, which will then call the notifyViewControllerDeallocated.

can you write a unit test for this? so that we don't break this unintentionally.

chunhtai commented on shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:
```diff
@@ -760,6 +764,47 @@ - (void)showAutocorrectionPromptRectForStart:(NSUInteger)start
+}
+
+- (void)scribbleInteractionBegan {
+  [_textInputChannel.get() invokeMethod:@"TextInputClient.scribbleInteractionBegan" arguments:nil];
```
should there be a framework pr associated with the change? it may be easier to review if you can also create the framework part pr

fbcouch replied:
Yes, I should have put a link to it in the description. Here it is: https://github.com/flutter/flutter/pull/75472

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:
```diff
@@ -471,6 +473,7 @@ - (void)setupChannels {
 
   _textInputPlugin.reset([[FlutterTextInputPlugin alloc] init]);
   _textInputPlugin.get().textInputDelegate = self;
+  [_textInputPlugin.get() setupIndirectScribbleInteraction:[self viewController]];
```
s/[self viewController]/self.viewController; here and everywhere else.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputDelegate.h:
```diff
@@ -38,6 +38,16 @@ typedef NS_ENUM(NSInteger, FlutterFloatingCursorDragState) {
+- (void)focusElement:(UIScribbleElementIdentifier)elementIdentifier
+             atPoint:(CGPoint)referencePoint
+              result:(id)callback;
+- (void)requestElementsInRect:(CGRect)rect result:(id)callback;
```
Why is this parameter `id`?  Please specify a specific type.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1608,6 +1908,96 @@ - (void)setTextInputEditingState:(NSDictionary*)state {
+
+#pragma mark - Methods related to Scribble support
+
+- (void)setupIndirectScribbleInteraction:(FlutterViewController*)viewController {
```
There already is a public setter for this via the property `setViewController`.  If people are expected to use this instead considering putting this logic in that method or making the property readonly.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -505,6 +557,9 @@ @implementation FlutterTextInputView {
   const char* _selectionAffinity;
   FlutterTextRange* _selectedTextRange;
   CGRect _cachedFirstRect;
+  NSArray* _selectionRects;
```
Use property instead of ivar so the memory management is explicit.  I don't see a release for this array so it's probably leaking.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1032,12 +1146,25 @@ - (void)setEditableTransform:(NSArray*)matrix {
+      [NSNumber numberWithFloat:[rect[3] floatValue]], [NSNumber numberWithInt:[rect[4] intValue]]
+    ]];
+  }
+  _selectionRects = rectsAsRect;
```
This is leaking, making it a property as I mentioned above will help you fix that.

gaaclarke replied:
You need to actually used this as a property to get the memory management benefits `self.selectionRects = rectsAsRect;`

fbcouch replied:
So, the issue is that this is the setter for that property – so if I reference it as `self.selectionRects`, I get an error about `_selectionRects` not being referenced:

```
../../flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:1182:1: error: ivar '_selectionRects' which backs the property is not referenced in this property's accessor [-Werror,-Wunused-property-ivar]```

So, it looks to me like `setSelectedTextRangeLocal` does a similar thing where it sets `_selectedTextRange` directly, so I'm not sure what the right thing to do is here – perhaps I just need to rename this method to not override the setter?

fbcouch replied:
Would adding `autorelease` to the `rectsAsRect` declaration help?

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1032,12 +1146,25 @@ - (void)setEditableTransform:(NSArray*)matrix {
+  for (NSUInteger i = 0; i < [rects count]; i++) {
+    NSArray* rect = rects[i];
+    [rectsAsRect addObject:@[
+      [NSNumber numberWithFloat:[rect[0] floatValue]],
```
It looks like you are unboxing and boxing a float value here, any reason why?

fbcouch replied:
This is probably the result of a few iterations of fumbling around with some of the various issues I've encountered – it seems to work fine without it, though, so I'll take that out

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1075,19 +1223,108 @@ - (CGRect)caretRectForPosition:(UITextPosition*)position {
-  NSUInteger currentIndex = ((FlutterTextPosition*)_selectedTextRange.start).index;
-  return [FlutterTextPosition positionWithIndex:currentIndex];
+  if ([_selectionRects count] == 0) {
+    NSUInteger currentIndex = ((FlutterTextPosition*)_selectedTextRange.start).index;
```
Add assert to guard cast.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1075,19 +1223,108 @@ - (CGRect)caretRectForPosition:(UITextPosition*)position {
 - (UITextPosition*)closestPositionToPoint:(CGPoint)point withinRange:(UITextRange*)range {
-  // TODO(cbracken) Implement.
-  return range.start;
+  NSUInteger start = ((FlutterTextPosition*)range.start).index;
```
Add assert for the casts.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1075,19 +1223,108 @@ - (CGRect)caretRectForPosition:(UITextPosition*)position {
+                     [_selectionRects[i][2] floatValue], [_selectionRects[i][3] floatValue]);
+      CGPoint pointForComparison =
+          CGPointMake(rect.origin.x, rect.origin.y + rect.size.height * 0.5);
+      float yDist = abs(pointForComparison.y - point.y);
```
s/abs/fabs;  You are sending a floating point number and want a floating point number.  If you are sure you have 32bit floats, use fabsf.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1075,19 +1223,108 @@ - (CGRect)caretRectForPosition:(UITextPosition*)position {
+          CGPointMake(rect.origin.x, rect.origin.y + rect.size.height * 0.5);
+      float yDist = abs(pointForComparison.y - point.y);
+      float xDist = abs(pointForComparison.x - point.x);
+      if (_closestIndex == 0 ||
```
nit: pull these boolean expressions into local variables so they are annotated with english.  Or split it into static function with a clear name.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1075,19 +1223,108 @@ - (CGRect)caretRectForPosition:(UITextPosition*)position {
+
+  FlutterTextRange* textRange = [[FlutterTextRange
+      rangeWithNSRange:fml::RangeForCharactersInRange(self.text, NSMakeRange(0, self.text.length))]
+      copy];
```
Why is there a copy here?

fbcouch replied:
No idea...I either copied that from somewhere else, or just added it for some reason I've entirely forgotten

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1075,19 +1223,108 @@ - (CGRect)caretRectForPosition:(UITextPosition*)position {
+                     [_selectionRects[i][2] floatValue], [_selectionRects[i][3] floatValue]);
+      CGPoint pointForComparison =
+          CGPointMake(rect.origin.x + rect.size.width, rect.origin.y + rect.size.height * 0.5);
+      float yDist = abs(pointForComparison.y - point.y);
```
fabs

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1075,19 +1223,108 @@ - (CGRect)caretRectForPosition:(UITextPosition*)position {
+          CGPointMake(rect.origin.x + rect.size.width, rect.origin.y + rect.size.height * 0.5);
+      float yDist = abs(pointForComparison.y - point.y);
+      float xDist = abs(pointForComparison.x - point.x);
+      if (yDist < _closestY - 1 ||
```
Looks like duplicated logic from above, more reason to make it a static function.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1150,10 +1387,52 @@ - (BOOL)hasText {
+- (UITextPlaceholder*)insertTextPlaceholderWithSize:(CGSize)size {
+  [_textInputDelegate insertTextPlaceholderWithSize:size withClient:_textInputClient];
+  _hasPlaceholder = YES;
+  return [[FlutterTextPlaceholder alloc] init];
```
You shouldn't return values that aren't released by the autorelease pool (read as this PR won't be accepted unless that change is made).

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1270,9 +1549,11 @@ @interface FlutterTextInputPlugin ()
 
 @implementation FlutterTextInputPlugin {
   NSTimer* _enableFlutterTextInputViewAccessibilityTimer;
+  NSMutableDictionary<UIScribbleElementIdentifier, NSValue*>* _scribbleElements;
```
use a property so the memory management is explicit

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1283,6 +1564,7 @@ - (instancetype)init {
     _autofillContext = [[NSMutableDictionary alloc] init];
     _activeView = [_reusableInputView retain];
     _inputHider = [[FlutterTextInputViewAccessibilityHider alloc] init];
+    _scribbleElements = [[NSMutableDictionary alloc] init];
```
You never release this so it's leaking.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1337,13 +1619,21 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
+  // TODO(fbcouch): only do this on iPadOS?
+  // This is necessary to set up where the scribble interactable element will be
+  _activeView.frame =
+      CGRectMake([dictionary[@"transform"][12] intValue], [dictionary[@"transform"][13] intValue],
```
12 and 13 should be annotated variables.

gaaclarke reviewed (changes requested):
I gave this a quick glance, seems promising thanks.  In its current state there is still a fair bit of slapdash memory management.  You're going to have to clean that up before merging.

chinmaygarde commented:
@fbcouch Any updates based on the Aarons review?

fbcouch commented:
@chinmaygarde I am hoping to address the new comments either tomorrow or early next week (thanks @gaaclarke for the feedback – I will readily admit I don't really know what I'm doing with memory management in objc, but I'll try to make the adjustments requested)

fbcouch commented:
@gaaclarke @chinmaygarde @chunhtai I believe I have made the necessary adjustments here. Let me know if there's anything else.

fbcouch commented:
@gaaclarke @chinmaygarde @chunhtai Just checking in here if there are any more changes needed. It looks like there are some merge conflicts now, so if everything looks good, I can clean those up and hopefully we can get this merged 😄

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.h:
```diff
@@ -8,11 +8,14 @@
 @interface FlutterTextInputPlugin : NSObject
 
 @property(nonatomic, assign) id<FlutterTextInputDelegate> textInputDelegate;
+@property(nonatomic, readonly) FlutterViewController* viewController;
```
This is a heavy import and providing a lot of access to the text input that it does't need.  Please look into how maybe textInputDelegate can solve your need or create a new protocol.

fbcouch replied:
Upon review, I think I may actually be able to get away with just `UIViewController`, so I will try that out.

fbcouch replied:
Yeah, that seems to work, so that's easy enough 😄

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.h:
```diff
@@ -23,6 +26,40 @@
 - (UIView<UITextInput>*)textInputView;
 
+// UIIndirectScribbleInteractionDelegate
+- (BOOL)indirectScribbleInteraction:(UIIndirectScribbleInteraction*)interaction
+                   isElementFocused:(UIScribbleElementIdentifier)elementIdentifier
+    API_AVAILABLE(ios(14.0));
```
Whey are these declared individually instead of implementing the protocol?

fbcouch replied:
The code highlighter does wonky things with the API_AVAILABLE annotation in the implementation file, but I will swap to that

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.h:
```diff
@@ -23,6 +26,40 @@
+ * These are used by the UIIndirectScribbleInteractionDelegate methods to handle focusing on the
+ * correct element
+ */
+- (void)setupIndirectScribbleInteraction:(FlutterViewController*)viewController;
```
Same comment about the FlutterViewController here.  Once you have a proper protocol you might want to make this a property since you want to set it and nil it out.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -445,6 +473,58 @@ - (UITextRange*)lineEnclosingPosition:(UITextPosition*)position {
+
+@implementation FlutterTextSelectionRect
+
+@synthesize rect = _rect;
+@synthesize writingDirection = _writingDirection;
+@synthesize containsStart = _containsStart;
+@synthesize containsEnd = _containsEnd;
+@synthesize isVertical = _isVertical;
```
I suspect these synthesize statements are superfluous.  You should only have to do this these days if you are implementing a protocol with properties.

fbcouch replied:
So, I believe the issue is that `FlutterTextSelectionRect` subclasses `UITextSelectionRect`, which has all of these properties as readonly. From a little bit of reading up on the errors that I get when removing these, I believe I either need this or to mark them as dynamic with e.g. `@dynamic rect;` and so on. It looks like `@dynamic` is used in the `FlutterViewController`, so I will swap to that, assuming that will be the preferred method.

fbcouch replied:
As it turns out, swapping this to dynamic breaks the tests - the setter selectors are missing in that case, apparently, so I think we do need the synthesize. This is a similar style to what's used in this file for the tokenizer and textInputDelegate.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -445,6 +473,58 @@ - (UITextRange*)lineEnclosingPosition:(UITextPosition*)position {
+                               containsStart:(BOOL)containsStart
+                                 containsEnd:(BOOL)containsEnd
+                                  isVertical:(BOOL)isVertical {
+  return [[FlutterTextSelectionRect alloc] initWithRectAndInfo:rect
```
Incorrect memory management, you need to return autoreleased objects from functions like this.

fbcouch replied:
Gotcha. Basically anywhere I'm calling `alloc`, I also need to `autorelease`, right? I tried to find all the places that might be in my changes.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -723,6 +852,13 @@ - (void)setSelectedTextRangeLocal:(UITextRange*)selectedTextRange {
   [self updateEditingState];
+  if (_scribbleInProgress || _scribbleFocused) {
+    _scribbleFocused = false;
+    FlutterTextRange* flutterTextRange = (FlutterTextRange*)selectedTextRange;
```
Add assert to cast.

fbcouch replied:
Done. I also found another place or two, so hopefully I've caught all of those (at least the ones in my changes).

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1062,10 +1212,31 @@ - (CGRect)firstRectForRange:(UITextRange*)range {
+      rangeWithNSRange:fml::RangeForCharactersInRange(self.text, NSMakeRange(0, self.text.length))]
+      copy];
+  for (NSUInteger i = 0; i < [_selectionRects count]; i++) {
+    if ([_selectionRects[i][4] unsignedIntegerValue] <= first &&
+        ((i + 1 == [_selectionRects count] && textRange.range.length > first) ||
+         (i + 1 < [_selectionRects count] &&
+          [_selectionRects[i + 1][4] unsignedIntegerValue] > first))) {
```
I recommend pulling this bool expression into a function or a local variable so people don't have to parse what it is trying to say.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.h:
```diff
@@ -74,8 +136,25 @@ FLUTTER_DARWIN_EXPORT
+@property(nonatomic, assign) FlutterViewController* viewController;
+@property(nonatomic) BOOL scribbleFocusing;
+@property(nonatomic) BOOL scribbleFocused;
+@property(nonatomic, assign) NSArray* selectionRects;
```
Please use generics for the NSArray.

gaaclarke replied:
`assign` is not what you are looking for here.

fbcouch replied:
I believe `strong` is the right thing to use here – updated

gaaclarke reviewed (commented):
I stopped reviewing halfway through because I saw a lot of memory management issues still.  Please give it another look.  This is good work, it just isn't up to the level where we could submit it.  Do you have a friend that has more iOS experience that could help you finish it up?

eseidelGoogle commented:
I just wanted to say thank you. An internal Google team approached me about this this morning and I was glad to see someone having already do so much work here.  Thank you @fbcouch

fbcouch commented:
@gaaclarke I addressed your comments and tried to apply those more broadly to the rest of the PR. I don't think I know anyone with quite the level of experience with memory management that's needed here...most of the people I know are more on the swift side of things 😄 

@eseidelGoogle Awesome! You're welcome, glad to contribute...flutter has been awesome to use, so it's cool to get to contribute something back 😄

fbcouch commented:
Also, I don't know what the deal is with that failing build_and_test_linux_unopt_debug check...it looks like something weird happened there, but I don't think that would be related to my changes

gaaclarke commented:
@fbcouch Yea that outstanding failure on CI shouldn't be related to your changes.  Once we have this approved you can rebase against master and hopefully it will clear up.

gaaclarke reviewed (changes requested):
Memory management is looking much better now, nice work.  There are a few things that you are doing that make the code a bit more difficult to maintain that should be addressed (like storing heterogeneous integers in _selectionRects instead of using structured data).

What is your plan for testing?  The test coverage of the unit tests is pretty spartan.  Is there going to be a higher level integration test?

(Heads up I'm going to be out for the next 2 weeks.)

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputDelegate.h:
```diff
@@ -38,6 +38,16 @@ typedef NS_ENUM(NSInteger, FlutterFloatingCursorDragState) {
                                          end:(NSUInteger)end
                                   withClient:(int)client;
+- (void)showToolbar:(int)client;
+- (void)focusElement:(UIScribbleElementIdentifier)elementIdentifier
+             atPoint:(CGPoint)referencePoint
+              result:(id)callback;
```
Please add the correct type here instead of using `id` as well.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1749,6 +2070,103 @@ - (void)setTextInputEditingState:(NSDictionary*)state {
+  _activeView.scribbleFocusing = YES;
+  [_textInputDelegate focusElement:elementIdentifier
+                           atPoint:focusReferencePoint
+                            result:^(id _Nullable result) {
```
(The FlutterTextInputPlugin is using the FlutterTextInputDelegate)

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1749,6 +2070,103 @@ - (void)setTextInputEditingState:(NSDictionary*)state {
+                         completion:
+                             (void (^)(NSArray<UIScribbleElementIdentifier>* elements))completion
+    API_AVAILABLE(ios(14.0)) {
+  [_textInputDelegate
```
(The FlutterTextInputPlugin is using the FlutterTextInputDelegate)

gaaclarke reviewed (commented):
I had a spare moment so I tried to do some of the cleanup for you.  One thing I noticed is that you are introducing for the first time when FlutterTextInputPlugin is calling into the FlutterTextInputDelegate.  Typically it is the FlutterTextInputView that is doing that (I would have actually named it the FlutterTextInputViewDelegate).  Is this intentional?  It seems like it would be better to match the established pattern of just letting the FlutterTextInputView call into that.

gaaclarke commented:
I tried to do some of the cleanup for you but I don't have permissions to push to this PR, feel free to grab it here: https://github.com/gaaclarke/engine/commit/4a6d5343d91d4e683ebc84d27c2b13b8769b0496

fbcouch commented:
@gaaclarke Thanks for the additional comments – I'll try to carve out some time to work on these over the next week or two

gaaclarke commented:
Hey, I'm back from vacation.  I saw there were some changes but another review wasn't requested so I'm just checking to see where you are with this.

fbcouch commented:
@gaaclarke Welcome back! I pulled in your commit and refactored out an interface for the UIIndirectScribbleInteraction stuff, I think the next steps are going to be creating an scribble focus status enum and a some kind of SelectionRect struct, then trying to beef up the test suite a little bit...speaking of, do you know if there's a simple way to run a single objc test? or just a single file, even? That would help me quite a bit with those – right now my flow is make a change -> run the entire objc test suite and scroll through the logs to find what I'm looking for.

gaaclarke commented:
Yea, if you open the xcode project associated with the test you can hit the button associated with the test to run just that one.  If you are doing the flutter engine objc unit tests there is a catch that all the tests don't show up un Xcode until you've run them all (because some tests live in shared libraries).  There is also the catch that if you change the code you have to run the build outside of Xcode before executing the tests again because it doesn't know how to compile them, it just links them.

Clear as mud?  I'll add a flag to run_tests.py, it shouldn't be that hard.

gaaclarke commented:
(I added the ability to run only one test to the objc unit tests too https://github.com/flutter/engine/pull/27566)

fbcouch commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1269,12 +1514,60 @@ - (BOOL)hasText {
+        _selectionRects[i][1],
+        _selectionRects[i][2],
+        _selectionRects[i][3],
+        [NSNumber numberWithInt:rectPosition],
```
Since we are already extending the `UITextSelectionRect` class as `FlutterTextSelectionRect`, I went ahead and added a `position` prop there and refactored the code accorditly

fbcouch commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.h:
```diff
@@ -10,9 +10,12 @@
+@interface FlutterTextInputPlugin : NSObject <UIIndirectScribbleInteractionDelegate>
 
 @property(nonatomic, assign) id<FlutterTextInputDelegate> textInputDelegate;
+@property(nonatomic, readonly) UIViewController* viewController;
```
`UIResponder` doesn't include the `view`, unfortunately, so I created an interface for `FlutterViewResponder` that exposes the view and the relevant touch methods

fbcouch commented:
@gaaclarke Thanks, the ability to run an individual test from the CLI was very helpful. I pulled in your commit, and I believe I've cleaned everything else up that you mentioned. I also added a few more tests, so hopefully the test suite is looking better now as well.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -355,6 +355,34 @@ static FlutterAutofillType autofillTypeOf(NSDictionary* configuration) {
+  float xDistOther = fabs(pointForOtherPosition.x - point.x);
+
+  BOOL isCloserVertically = yDist < yDistOther - 1;
+  BOOL isEqualVertically = yDist >= yDistOther - 1 && yDist <= yDistOther + 1;
```
You might want to factor this out to a function `isApproximately(x, y, delta)`

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -355,6 +355,34 @@ static FlutterAutofillType autofillTypeOf(NSDictionary* configuration) {
   return FlutterAutofillTypeNone;
 }
 
+static BOOL isPositionCloserToPoint(CGPoint point,
```
Please, add docstring.  I'm not sure what this function should do when just reading the function signature.   I think you are asking is `point` closer to `position` or `otherPosition`?  We should rename position to match the datatype better since position seems to imply a point, how about "region"?

You seem to be doing distance math but you aren't using the distance formula.  If you want to see if a point is closer to one rect or another I'd just use the distance formula against the 4 corners.  (you could omit the sqrt for speed).

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -751,11 +870,12 @@ - (void)pressesBegan:(NSSet<UIPress*>*)presses
-                                }];
+      [_textInputDelegate
+          flutterTextInputView:self
+              handlePressEvent:[[FlutterUIPressProxy alloc] initWithPress:press withEvent:event]
```
This is a leak, FlutterUIPressProxy should be released here or autoreleased.

fbcouch replied:
Ah...this is actually something new that was introduced when I merged in master last time...I updated it to pass in the flutterTextInputView, but I'll go ahead and add these to the autorelease pool as well.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -767,10 +887,11 @@ - (void)pressesChanged:(NSSet<UIPress*>*)presses
-                  [super pressesChanged:[NSSet setWithObject:press] withEvent:event];
-                }];
+          flutterTextInputView:self
+              handlePressEvent:[[FlutterUIPressProxy alloc] initWithPress:press withEvent:event]
```
Same with this PressProxy.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -781,11 +902,12 @@ - (void)pressesEnded:(NSSet<UIPress*>*)presses
-                                }];
+      [_textInputDelegate
+          flutterTextInputView:self
+              handlePressEvent:[[FlutterUIPressProxy alloc] initWithPress:press withEvent:event]
```
Same here

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -797,10 +919,11 @@ - (void)pressesCancelled:(NSSet<UIPress*>*)presses
-                  [super pressesCancelled:[NSSet setWithObject:press] withEvent:event];
-                }];
+          flutterTextInputView:self
+              handlePressEvent:[[FlutterUIPressProxy alloc] initWithPress:press withEvent:event]
```
Same here

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -991,6 +1151,9 @@ - (void)setMarkedText:(NSString*)markedText selectedRange:(NSRange)markedSelecte
   NSRange selectedRange = _selectedTextRange.range;
   NSRange markedTextRange = ((FlutterTextRange*)self.markedTextRange).range;
 
+  if (_scribbleInProgress || _scribbleFocusStatus != FlutterScribbleStatusUnfocused)
```
You need curly braces here (https://google.github.io/styleguide/objcguide.html#conditionals)

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1040,12 +1203,11 @@ - (UITextRange*)textRangeFromPosition:(UITextPosition*)fromPosition
 
 - (NSUInteger)decrementOffsetPosition:(NSUInteger)position {
-  return fml::RangeForCharacterAtIndex(self.text, MAX(0, position - 1)).location;
+  return MAX(0, position - 1);
```
Why was this removed?  This looks like it might be breaking UTF stuff?

fbcouch replied:
Well, it's actually needed for scribble to handle UTF stuff properly – for whatever reason, UIKit seems to want to inspect every position (even within a multi-byte character sequence) in order to function properly. I'll see about wrapping that in a check for `_scribbleInProgress` to make this more conservative

fbcouch replied:
Turns out that my _scribbleInProgress check wasn't working – I don't need this due to the short circuit in `- positionFromPosition`

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1250,19 +1436,88 @@ - (CGRect)bounds {
+    NSUInteger position = _selectionRects[i].position;
+    if (position >= start && position <= end) {
+      BOOL isFirst = _closestIndex == 0;
+      if (isFirst || isPositionCloserToPoint(point, _selectionRects[i].rect, _closestRect, NO)) {
```
nit: Adding a comment here would be helpful for the `NO` argument, `/*checkRightBoundary=*/ NO`

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterViewResponder.h:
```diff
@@ -0,0 +1,22 @@
+#define SHELL_PLATFORM_IOS_FRAMEWORK_SOURCE_FLUTTERVIEWRESPONDER_H_
+
+#import <Foundation/Foundation.h>
+
```
This too should have NS_ASSUME_NONNULL_BEGIN.

gaaclarke reviewed (changes requested):
Hey, we are making good progress with this PR, getting really close now.  There is a lot of code.  I want to go back and double check your code coverage.  What's the easiest way to test this with the simulator?

fbcouch commented:
@gaaclarke Great! I'll take a look at these. Unfortunately, I don't believe there's a way to test it with the simulator – as far as I'm aware, all pencil stuff has to be done with physical hardware.

fbcouch commented:
@gaaclarke I think I've got everything from the previous review. There is one last issue that I need to get figured out – maybe you will have some insight – for some reason, the OS seems to be adding a grey shaded rectangle to the screen when using scribble to delete. I suspect it is related to autocorrect highlighting, but I'm not sure. So far I haven't been able to get rid of it. It does go away when tapping the screen again.

I _think_ that behavior is new-ish, so I'm going to try to go back and check previous merges from master into this branch to see if I can track down where it was introduced.

gaaclarke commented:
Let me get my hands on a device, if we are feeling good about the code I can help debug.  We're all still working from home so my device situation is a bit off.

fbcouch commented:
Sounds good. I think we're in good shape other than this issue and one other that I came across.

I traced this issue down to the addition of a UITextInteraction (in the init method for the FlutterTextInputView). I've tried removing it temporarily for the Scribble interaction, then adding it back, and that seems to work for a bit, but the issues come back eventually (usually after swapping focus to another TextField). I've tried removing the UITextSelectionView subview (added by UITextInteraction) and setting it's frame to CGRectZero, and neither of those seems to have any effect. Probably the next thing to try is messing with the UIPanGestureRecognizer that the UITextInteraction adds.

The other issue that may be related (need to do more testing) is that tapping or double tapping words in the TextView seems to trigger the native suggestion/spell checking menu.

gaaclarke commented:
I tried to test this locally with this PR and the Flutter PR and I get the following error:
```
[  +11 ms] executing: [/Users/aaclarke/dev/flutter-examples/scribble/ios/] xcrun xcodebuild -list
[   +9 ms] <- compile package:scribble/main.dart
[  +90 ms] ERROR: FormatException: Could not find an option named "flutter-widget-cache".
[   +1 ms] Usage: server [options] [input.dart]
```

Are both of the PRs in sync with eachother and up to date?  For `flutter` I'm at 1b6f950992304ea3c7bee832b0af4b274595d4d0 and for the engine I'm at 0ad64968b9ff2520da920d754a91f54bf8682cff.

fbcouch commented:
@gaaclarke Can you try https://github.com/flutter/engine/pull/24224/commits/a95f2d9cb6e20979043e2f99af65fa39319c36c1 for the engine? That should match this branch (twinsunllc:feature/scribble+master) – I think the one you are on is the branch we are currently using in our app (twinsunllc:feature/scribble), which was started from flutter 1.22.2, and actually includes another change that I'll probably try to create a PR for later (enabling undo support).

The flutter commit looks correct, though.

I pushed a workaround in this latest commit, but I'm not super happy with it...moving between input fields sometimes ends up in a weird state where they are un-tappable. That might be related to filtering out too many gestures. I'll have to look at that some more.

fbcouch commented:
I pushed up a new workaround that I'm happier with – instead of filtering out gestures, I delayed the setup of the UITextInteraction to the `- beginFloatingCursorAtPoint:` method, and remove it in `- endFloatingCursor:`. I left the overrides of the `- insertionPointColor`, `- selectionBarColor`, and `- selectionHighlightColor` to keep the native selection stuff hidden. I also moved where the UIScribbleInteraction is set up to make that more reliable – that was causing some issues as well.

@LongCatIsLooong I think you added the UITextInteraction bit – do you think delaying the setup to `- beginFloatingCursorAtPoint:` will cause any issues, like bringing back the original issue that you added it to prevent?

I may take another look at leaving the setup in `- init` and adding/removing the interaction in the scribble delegate hooks. I think the issues I was having with the UIScribbleInteraction not getting set up were probably causing the issues I had with that attempt.

fbcouch commented:
All right, I tried to reproduce #70267 with or without the UITextInteraction, and wasn't able to in either case, so it would be helpful if someone else could make an attempt just to make sure. I did find out that changing the keyboard language to Japanese would crash the app due to the NSAssert in selectionRectsForRange checking that the text range contains FlutterTextPositions, so I added a guard clause to return early in that case.

Other than that, I'm pretty happy with the state this is in, so let me know if there are any more issues that you see with the code or run into when testing.

fbcouch commented:
Looks like merging in master fixed up the checks that were failing (except the luci-engine one)... @gaaclarke or @LongCatIsLooong any additional thoughts on this one?

jmagman commented:
Friendly ping, @gaaclarke.

justinmc commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.h:
```diff
@@ -24,6 +42,13 @@
 
+/**
+ * These are used by the UIIndirectScribbleInteractionDelegate methods to handle focusing on the
+ * correct element
```
Nit: Period here.

justinmc commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -355,6 +355,61 @@ static FlutterAutofillType autofillTypeOf(NSDictionary* configuration) {
+  float yDistOther = fabs(pointForOtherSelectionRect.y - point.y);
+  float xDistOther = fabs(pointForOtherSelectionRect.x - point.x);
+
+  BOOL isCloserVertically = yDist < yDistOther - 1;
```
Why the `- 1` here?

fbcouch replied:
It's been awhile, but I think the idea was similar to `isApproximatelyEqual`, where there's a little bit of preference in favor of the `otherSelectionRect` to avoid unnecessary churn. I think where this sort of thing comes in most handy is when you have an emoji in the string, sometimes the center point of those is offset from the rest of the line a little bit.

justinmc replied:
Maybe consider adding a comment explaining that.

justinmc commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -355,6 +355,61 @@ static FlutterAutofillType autofillTypeOf(NSDictionary* configuration) {
+// If checkRightBoundary is set, the right-center point on selectionRect and
+// otherSelectionRect will be used instead of the left-center point.
+//
+// This uses special (empirically determined) logic for determining the closer rect,
+// rather than a simple distance calculation. First, the closer vertical distance is
+// determined. Within the closest y distance, if the point is above the bottom
+// of the closest rect, the x distance will be minimized; however, if the point is
+// below the bottom of the rect, the x value will be maximized.
```
It sounds like you've tried this natively on an iPad and this is what you've observed the behavior to be, rather than just a pure distance calculation?  If that's true, consider adding the exact device and iOS version that you used to determine this behavior to this comment.  Maybe `empirically determined on an ... running ...`.  That way if someone in the future notices that it doesn't perfectly match native, they have some idea of where this algorithm came from.

justinmc commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -355,6 +355,61 @@ static FlutterAutofillType autofillTypeOf(NSDictionary* configuration) {
+                                 (isBelowBottomOfLine && isFartherToRight))));
+}
+
+// Checks whether Scribble features are possibly available – meaning this is an iPad running iOS 14+
```
Nit: period.

justinmc commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -355,6 +355,61 @@ static FlutterAutofillType autofillTypeOf(NSDictionary* configuration) {
+static BOOL isScribbleAvailable() {
+  if (@available(iOS 14.0, *)) {
+    NSString* deviceModel = (NSString*)[UIDevice currentDevice].model;
+    if ([[deviceModel substringWithRange:NSMakeRange(0, 4)] isEqualToString:@"iPad"]) {
```
I was wondering if we had a better way to do this, but I couldn't find one.  I didn't see anywhere else in the engine where we try to detect iPad.  I guess this is good enough.

gaaclarke replied:
`if ( UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad )` is another way

justinmc commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -732,6 +870,28 @@ - (void)setTextInputState:(NSDictionary*)state {
   }
 }
 
+// Forward touches to the viewResponder to allow tapping inside the UITextField as normal
```
Nit: period.

justinmc commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -839,9 +1007,35 @@ - (BOOL)isVisibleToAutofill {
 // account.
 - (void)setIsVisibleToAutofill:(BOOL)isVisibleToAutofill {
+  // This probably needs to change (think it is getting overwritten by the updateSizeAndTransform
+  // stuff for now)
```
Nit: period.

justinmc commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -852,6 +1046,22 @@ - (BOOL)canBecomeFirstResponder {
 
+- (BOOL)canPerformAction:(SEL)action withSender:(id)sender {
+  // When scribble is available, the FlutterTextInputView will display the native toolbar unless
+  // these text editing actions are disabled
```
Nit: period.

justinmc commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1546,13 +1947,28 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
 - (void)setEditableSizeAndTransform:(NSDictionary*)dictionary {
   [_activeView setEditableTransform:dictionary[@"transform"]];
+  if (isScribbleAvailable()) {
+    // This is necessary to set up where the scribble interactable element will be
```
Nit: period.

justinmc reviewed (commented):
Just some small stuff/nits as I had been meaning to take a look after reviewing the framework PR.  I'll defer to @gaaclarke for approval.

justinmc commented:
@fbcouch FYI there are some conflicts now.

fbcouch commented:
@justinmc This one should be ready to go now

justinmc commented:
@fbcouch Looks like another conflict popped up.

gaaclarke commented:
Okay so I tested it locally.  It mostly works.  Drawing a line through text to select it works, but the highlight of the region isn't rendered:
![IMG_0005](https://user-images.githubusercontent.com/30870216/139482561-a6e5e11c-a39e-4141-99cc-3279b4f5df29.PNG)

If I background the app, then bring it back, the text is selected correctly.

gaaclarke commented:
Hey @justinmc since this mostly works, do you think we can land this behind a flag?  So @fbcouch doesn't have to maintain these 2 big PR's?  Then they could just land little patches to fix things up and when we are happy with it we can remove the flag?

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -397,6 +397,65 @@ static FlutterAutofillType autofillTypeOf(NSDictionary* configuration) {
 }
 
+static BOOL isApproximatelyEqual(float x, float y, float delta) {
+  return x >= y - delta && x <= y + delta;
```
fyi: `fabsf(x - y) <= delta` is less operations (3-5 versus 3)

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -625,14 +753,11 @@ - (instancetype)init {
       _smartQuotesType = UITextSmartQuotesTypeYes;
       _smartDashesType = UITextSmartDashesTypeYes;
     }
+    _selectionRects = @[];
```
This is a potential crash, it should be `_selectionRects = [@[] retain]`;  or `_selectionRects = [[NSArray alloc] init]`.  The literal is on the autorelease pool and can be deleted before you access it.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -900,6 +1094,22 @@ - (BOOL)canBecomeFirstResponder {
+       action == @selector(delete:) || action == @selector(makeTextWritingDirectionLeftToRight:) ||
+       action == @selector(makeTextWritingDirectionRightToLeft:) ||
+       action == @selector(toggleBoldface:) || action == @selector(toggleItalics:) ||
+       action == @selector(toggleUnderline:) || action == NSSelectorFromString(@"_define:") ||
```
`_define:` is a private selector?  This may get flagged by iOS app review.

fbcouch replied:
That's a good point...we might have to make a whitelist instead of a blacklist to get around that...I'll look into that.

fbcouch replied:
Actually, I think we can just return NO for everything – I _think_ the big thing I was worried about was saying no to the accessibility ones, but I just tried it with VoiceOver and it seems to still be able to navigate/read the TextInput.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1478,20 +1830,63 @@ - (void)updateEditingStateWithDelta:(FlutterTextEditingDelta*)delta {
+
+  _scribbleFocusStatus = FlutterScribbleFocusStatusUnfocused;
+  [self resetScribbleInteractionStatusIfEnding];
+  _selectionRects = copiedRects;
```
This is leaking the old value of `_selectionRects`, use `self.selectionRects = copiedRects; [copiedRects release];`

justinmc commented:
I'm on board with merging this behind a flag 👍

fbcouch commented:
@gaaclarke Thanks for the review and testing it out. This is interesting, because it seems that a MaterialTextField works correctly, but CupertinoTextField doesn't. I'll dig into that, but do you or @justinmc know off the top of your head if there's some difference in how they display that highlight?

I like the idea of landing behind a flag – how difficult will that be to enable in an app? (We're currently frozen on flutter 1.22.6 to avoid having to make multiple custom engine versions)

fbcouch commented:
So the issue with the CupertinoTextField appears to be that focusing with the pencil doesn't trigger the `_handleFocusChanged` callback, so the widget doesn't get rebuilt knowing that it has focus. I'll see if I can figure out a quick fix for that.

fbcouch commented:
Yep, it wasn't adding the focus listener in initState like the Material version does (https://github.com/flutter/flutter/pull/75472/commits/f30930211fe0e9fdf7fbf591c9378672a655c9d9). So, @gaaclarke could you test again? Thanks!

gaaclarke reviewed (approved):
I tested it out and it works for me.  I looked through the code and it looks good now.  I don't see any obvious errors.  Thanks Jami, awesome job!

fbcouch commented:
@gaaclarke Awesome, thanks for all your time giving feedback on this!

morio77 commented:
@fbcouch 

I am using the flutter master channel.

Is it a bug that when I draw with Apple Pencil with scribble turned on in iPad settings, scribble trails are displayed in most widgets including AppBar widget?

Also, is this related to this PR?

By the way, the problem does not occur in flutter 2.0.0 to 2.8.1, only in the master channel.

fbcouch commented:
@morio77 I think that is an issue that will be resolved once the framework PR is merged (see flutter/flutter#75472 ). If you use that branch instead, things should work properly. I think the issue is that the `TextInputClient.focusElement` method invocation is not being responded to properly by the framework, so the `completion` method is not getting called in a timely manner in the `- focusElementIfNeeded` method

cyanglaz commented on shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:
```diff
@@ -331,6 +334,7 @@ - (void)setViewController:(FlutterViewController*)viewController {
 
 - (void)attachView {
   self.iosPlatformView->attachView();
+  [_textInputPlugin.get() setupIndirectScribbleInteraction:self.viewController];
```
Came across this while looking into refactoring `attachView`. Does it make sense to move this line somewhere else? `setupIndirectScribbleInteraction:self` doesn't sound like should be a part of "attachView"

fbcouch replied:
Maybe there's a better place for it...iirc it's purpose here is to make sure the indirect scribble interaction is set up along with the view – that is something that needs to always be there so that we can capture scribble interactions that happen outside of text areas
